extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs

```diff
@@ -57,136 +59,182 @@ private async ValueTask<ConversionResult> ConvertFromBindingDataAsync(ConverterC
-                    @"Binding parameters to complex objects uses JSON serialization.
-                    1. Bind the parameter type as 'string' instead to get the raw values and avoid JSON deserialization, or
-                    2. Change the blob to be valid json.
-                    The JSON parser failed: {0}",
-                    ex.Message);
+                if (targetType == typeof(BlobContainerClient) && BlobIsFileRegex.IsMatch(blobData.BlobName))
+                {
+                    throw new InvalidOperationException("Binding to a BlobContainerClient with a blob path is not supported. "
+                                                        + "Either bind to the container path, or use BlobClient instead.");
+                }
 
-                return ConversionResult.Failed(new InvalidOperationException(msg, ex));
-            }
-            catch (Exception ex)
-            {
-                return ConversionResult.Failed(ex);
+                if (targetType != typeof(BlobContainerClient) && string.IsNullOrEmpty(blobData.BlobName))
+                {
+                    throw new InvalidOperationException($"'{nameof(blobData.BlobName)}' cannot be null or empty when binding to a single blob.");
+                }
+
+                return await ToTargetTypeAsync(targetType, container, blobData.BlobName!);
             }
         }
 
-        private Dictionary<string, string> GetBindingDataContent(ModelBindingData bindingData)
+        private bool IsCollectionBinding(Type targetType, string blobName)
         {
-            return bindingData?.ContentType switch
+            // Edge case: These two types should be treated as a single blob binding
+            // string implements IEnumerable<char> and byte[] would pass the IsArray check
+            if (targetType == typeof(string) || targetType == typeof(byte[]))
             {
-                Constants.JsonContentType => new Dictionary<string, string>(bindingData?.Content?.ToObjectFromJson<Dictionary<string, string>>(), StringComparer.OrdinalIgnoreCase),
-                _ => throw new InvalidContentTypeException(Constants.JsonContentType)
-            };
+                return false;
+            }
+
+            if (!(targetType.IsArray || typeof(IEnumerable).IsAssignableFrom(targetType)))
+            {
+                return false;
+            }
+
+            return true;
         }
 
-        private async Task<object?> ConvertModelBindingDataAsync(IDictionary<string, string> content, Type targetType, ModelBindingData bindingData)
+        private async Task<object> BindToCollectionAsync(Type targetType, Type elementType, BlobContainerClient container, string blobPath)
         {
-            content.TryGetValue(Constants.Connection, out var connectionName);
-            content.TryGetValue(Constants.ContainerName, out var containerName);
-            content.TryGetValue(Constants.BlobName, out var blobName);
+            if (BlobIsFileRegex.IsMatch(blobPath))
```

Comment:
Would it make sense to move this check to the `IsCollectionBinding` method, returning false if it's a file and causing the logic to fallback to single blob binding, which would remove the need to call `DeserializeToTargetObjectAsync` from multiple locations, avoiding duplication and the special casing in this path.

Reply:
Ooo good call out!! I didn't go for that initially because I thought it would end up hitting the (for example) string single case, but we don't have a switch case for `string[]` 🤦🏻‍♀️ so this would hit DeserializeToTargetObjectAsync and work as expected - will fix :)